src/Common/Contracts/Container.php

```diff
@@ -27,4 +27,89 @@ public function get( $id ) {
 			throw new Not_Bound_Exception( $e->getMessage(), $e->getCode() );
 		}
 	}
+
+	/**
+	 * Overrides the parent method to fire an action when a service provider is registered.
+	 *
+	 * @since TBD
+	 *
+	 * @param string $serviceProviderClass The service provider class name.
+	 * @param string ...$alias             Optional. The alias(es) to register the service provider with.
+	 *
+	 * @return void
+	 *
+	 * @throws \TEC\Common\lucatume\DI52\ContainerException If the provider class is marked as deferred but
+	 *                                                      does not provide a set of deferred registrations.
+	 */
+	public function register( $serviceProviderClass, ...$alias ) {
+		// Register the provider with the parent container.
+		parent::register( $serviceProviderClass, ...$alias );
+
+		/**
+		 * Fires when a service provider is registered by the container.
+		 *
+		 * @since TBD
+		 *
+		 * @param string        $serviceProviderClass The service provider class name.
+		 * @param array<string> $alias                The alias(es) the service provider was registered with.
+		 */
+		do_action( 'tec_container_registered_provider', $serviceProviderClass, $alias );
+
+		/**
+		 * Fires a class-specific action when a service provider is registered by the container.
+		 *
+		 * @since TBD
+		 *
+		 * @param array<string> $alias The alias(es) the service provider was registered with.
+		 */
+		do_action( 'tec_container_registered_provider_' . $serviceProviderClass, $alias );
+
+		if (
+			// Back compat with older definition of Service Provider.
+			! property_exists( $serviceProviderClass, 'registration_action' )
+			// New definition of Service Provider: default action is empty.
+			|| empty( $serviceProviderClass::$registration_action )
+		) {
+			return;
+		}
+
+		/**
+		 * Fires a custom action defined by the Service Provider when it's registered.
+		 *
+		 * @since TBD
+		 */
+		do_action( $serviceProviderClass::$registration_action, $serviceProviderClass, $alias );
```

Comment:
Quick question: Will there be any situation where `$serviceProviderClass::$registration_action` might not exist or do all service providers have it defined?